### PR TITLE
Add exclusion zones support

### PR DESCRIPTION
**Pull request by vdbg**

### CODE REVIEW

File: geofence.py [hunk 1]
```diff
@@ -35,52 +38,103 @@ def Y(self) -> float:
...
 
 
 class PolygonFence(Geofence):
-    def __init__(self, name: str, conf: dict, all_tiles: set[str], hubitatIds: set[int], hubitat_devices: dict[int, str]) -> None:
-        super().__init__(name, conf, all_tiles, hubitatIds, hubitat_devices)
+    def __init__(self, name: str, conf: dict, geoconf: GeoConfig, exclusion: bool) -> None:
+        super().__init__(name, conf, geoconf, exclusion)
 
         vertices: list[list[float]] = conf.get("vertices", [])
+        hashes: set[int] = set()
```
Comment (Nickolaim) [inline geofence.py]: why not set of Points? The requirement for set is the class to be hashable, that's not hard to do - https://docs.python.org/3/glossary.html#term-hashable

File: geofence.py [hunk 2]
```diff
@@ -89,8 +143,14 @@ def __init__(self, name: str, conf: dict, all_tiles: set[str], hubitatIds: set[i
 
         for data in vertices:
             if len(data) != 2:
-                raise Exception(f"Invalid (lat,long) pair: {data}")
-            self.p.append(Point(data[0], data[1]))
+                raise Exception(f"Invalid (lat,long) pair: {data}.")
+            point = Point(data[0], data[1])
+            hash_value = hash(point.latitude) ^ hash(point.longitude)
+            if hash_value in hashes:
+                raise Exception(f"Vertex {str(point)} used twice in '{self.name}'.")
```
Comment (Nickolaim) [inline geofence.py]: del str?

```
$ python3
Python 3.9.9 (main, Nov 21 2021, 03:23:44)
[Clang 13.0.0 (clang-1300.0.29.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> class Foo:
...     def __str__(self) -> str:
...       return "myfoo"
...
>>> print(f"{Foo()}")
myfoo
>>> print(f"{str(Foo())}")
myfoo
```

File: Dockerfile [hunk 3]
```diff
@@ -16,10 +16,13 @@ WORKDIR /app
...
 
+# Install location of upgraded pip
+ENV PATH /home/hubitile/.local/bin:$PATH
+
 COPY requirements.txt     /app
 
-RUN pip install --no-cache-dir --upgrade pip --no-warn-script-location && \
-    pip install --no-cache-dir -r ./requirements.txt --no-warn-script-location 
+RUN pip install --no-cache-dir --disable-pip-version-check --upgrade pip && \
+    pip install --no-cache-dir -r ./requirements.txt
```
Comment (Nickolaim) [inline Dockerfile:25]: Any reason why not to have 2 `RUN` statements?  It would look cleaner

File: geofence.py [hunk 4]
```diff
@@ -35,52 +38,100 @@ def Y(self) -> float:
...
     def __str__(self) -> str:
-        return f"(lat={self.latitude}.long={self.longitude};x={self.X()},y={self.Y()})"
+        return f"(lat={self.latitude};long={self.longitude})"
+
+
+class TileWrapper:
+    def __init__(self, tile: Tile):
+        self.tile = tile
+        self.fullname = f"'{self.tile.name}' ({self.tile.uuid})"
+        self.previouslyIgnored: bool = False
```
Comment (Nickolaim) [inline geofence.py]: `previously_ignored`?

File: geofence.py [hunk 5]
```diff
@@ -35,52 +38,100 @@ def Y(self) -> float:
...
 
     def __str__(self) -> str:
-        return f"(lat={self.latitude}.long={self.longitude};x={self.X()},y={self.Y()})"
+        return f"(lat={self.latitude};long={self.longitude})"
+
+
+class TileWrapper:
+    def __init__(self, tile: Tile):
+        self.tile = tile
+        self.fullname = f"'{self.tile.name}' ({self.tile.uuid})"
```
Comment (Nickolaim) [inline geofence.py:53]: name and uuid are properties, why not fullname?

File: geofence.py [hunk 6]
```diff
@@ -35,52 +38,100 @@ def Y(self) -> float:
...
+        key = "tiles"
 
-        if "tile_to_hubitat" not in conf:
-            logging.warn(f"No tiles or Hubitat devices associated with section '{name}'.")
+        if key not in conf:
+            logging.warn(f"No tiles associated with location '{name}'.")
             return
 
-        self.tiles = conf["tile_to_hubitat"]
+        self.tiles = {x: None for x in conf[key]} if exclusion else conf[key]
```
Comment (Nickolaim) [inline geofence.py:102]: This one is a little bit easier to read IMO: `{x: None if exclude else y for x, y in conf[key].items()}`

File: geofence.py [hunk 7]
```diff
@@ -35,52 +38,100 @@ def Y(self) -> float:
...
 
         for tile, hubitatId in self.tiles.items():
-            all_tiles.add(tile)
+            geoconf.all_tiles.add(tile)
+            if not hubitatId:
```
Comment (Nickolaim) [inline geofence.py]: Not sure if `hubitatId` can ever be `0`.  If yes, this code won't work.  Safer alternative is `if hubitatId is None:`

File: geofence.py [hunk 8]
```diff
@@ -35,52 +38,100 @@ def Y(self) -> float:
...
+                continue
             hubitatId = int(hubitatId)
-            if hubitatId in hubitatIds:
-                raise Exception(f"Hubitat device Id {hubitatId} is referenced in section '{name}' and another section.")
-            hubitatIds.add(hubitatId)
-            if hubitatId not in hubitat_devices:
+            if hubitatId in geoconf.hubitatIds:
+                raise Exception(f"Hubitat device Id {hubitatId} is referenced in location '{name}' and another location.")
+            geoconf.hubitatIds.add(hubitatId)
+            if not hubitatId in geoconf.hubitat_devices:
```
Comment (Nickolaim) [inline geofence.py]: what about this `if hubitatId not in geoconf.hubitat_devices` ?

File: geofence.py [hunk 9]
```diff
@@ -35,52 +38,100 @@ def Y(self) -> float:
...
                 raise Exception(f"Hubitat device Id {hubitatId} is not a virtual presence sensor exported by Hubitat's MakerAPI.")
 
     def isInside(self, p: Point) -> bool:
         return False
 
-    def processTile(self, p: Point, name: str, uuid: str, hubitat: Hubitat) -> None:
+    def processTile(self, tile: TileWrapper, hubitat: Hubitat) -> bool:
```
Comment (Nickolaim) [inline geofence.py:121]: Having `tile` as `TileWrapper` is weird.  Consider:
- `tile_wrapper`
- rename  `Tile` to `TileBase`, `TileWrapper` to `Tile`

File: geofence.py (hunk 2 above)
Comment (vdbg) [reply geofence.py]: duh. yes. thx

File: geofence.py (hunk 4 above)
Comment (vdbg) [reply geofence.py]: yes

File: geofence.py (hunk 5 above)
Comment (vdbg) [reply geofence.py:53]: replaced with str

File: geofence.py (hunk 6 above)
Comment (vdbg) [reply geofence.py:102]: not sure this works? in one case conf[key] is a list, in the other case it's a dictionary

File: geofence.py (hunk 8 above)
Comment (vdbg) [reply geofence.py]: what's the difference? just curious

File: geofence.py (hunk 9 above)
Comment (vdbg) [reply geofence.py:121]: Unfortunately, Tile is coming from the pytile package.

File: geofence.py (hunk 1 above)
Comment (vdbg) [reply geofence.py]: Is that the proposal? If so, not convinced it's a better alternative
```
--- a/geofence.py
+++ b/geofence.py
@@ -40,6 +40,12 @@ class Point:
     def __str__(self) -> str:
         return f"(lat={self.latitude};long={self.longitude})"

+    def __eq__(self, other) -> bool:
+        return self.longitude == other.longitude and self.latitude == other.latitude
+
+    def __hash__(self) -> int:
+        return hash(self.latitude) ^ hash(self.longitude)
+

 class TileWrapper:
     def __init__(self, tile: Tile):
@@ -137,7 +143,7 @@ class PolygonFence(Geofence):
         super().__init__(name, conf, geoconf, exclusion)

         vertices: list[list[float]] = conf.get("vertices", [])
-        hashes: set[int] = set()
+        hashes: set[Point] = set()

         if len(vertices) < 3:
             raise Exception(f"Polygon fence '{self}' needs at least 3 vertices.")
@@ -148,10 +154,9 @@ class PolygonFence(Geofence):
             if len(data) != 2:
                 raise Exception(f"Invalid (lat,long) pair: {data}.")
             point = Point(data[0], data[1])
-            hash_value = hash(point.latitude) ^ hash(point.longitude)
-            if hash_value in hashes:
+            if point in hashes:
                 raise Exception(f"Vertex {point} used twice in '{self}'.")
-            hashes.add(hash_value)
+            hashes.add(point)

             self.p.append(point)
```

Comment (Nickolaim) [reply geofence.py]: I am not sure if `__eq__()` is required.  But yes, that's mostly it.  With this change one incapsulate hash function in Point class.

File: geofence.py (hunk 6 above)
Comment (Nickolaim) [reply geofence.py:102]: Ooops.  I did not realize that the value of conf[key] can be a different type.  Consider using one type.  This is 100% WTF moment a year from now while debugging.

File: geofence.py (hunk 8 above)
Comment (Nickolaim) [reply geofence.py]: `if something not in a jar` this is how humans talk
`if not something in a jar` this is how computers talk

Computers talk weird.

File: geofence.py (hunk 9 above)
Comment (Nickolaim) [reply geofence.py:121]: ok!

File: geofence.py (hunk 7 above)
Comment (vdbg) [reply geofence.py]: ok

File: geofence.py (hunk 1 above)
Comment (vdbg) [reply geofence.py]: `__eq__()` is required.
https://hynek.me/articles/hashes-and-equality

> To make a class hashable, it has to implement both the __hash__(self) method and the aforementioned __eq__(self, other) method

File: geofence.py (hunk 6 above)
Comment (vdbg) [reply geofence.py:102]: won't fixing: hubitatIds only make sense when reporting presence

File: geofence.py (hunk 8 above)
Comment (vdbg) [reply geofence.py]: ok

File: Dockerfile (hunk 3 above)
Comment (vdbg) [reply Dockerfile:25]: It's logically one unit: install the python dependencies

File: geofence.py (hunk 1 above)
Comment (vdbg) [reply geofence.py]: ended up keeping the changes since I went through the trouble of writing them ;)

**Pull request merged**